xarray/core/resample.py

```diff
@@ -0,0 +1,342 @@
+           over this dimension.
+        2. Otherwise, stack over the new dimension given by name of this
+           grouping (the argument to the `groupby` function).
+
+        Parameters
+        ----------
+        func : function
+            Callable to apply to each sub-dataset.
+        **kwargs
+            Used to call `func(ds, **kwargs)` for each sub-dataset `ar`.
+
+        Returns
+        -------
+        applied : Dataset or DataArray
+            The result of splitting, applying and combining this dataset.
+        """
+        kwargs.pop('shortcut', None)  # ignore shortcut if set (for now)
+        applied = (func(ds, **kwargs) for ds in self._iter_grouped())
+        combined = self._combine(applied)
+
+        return combined.rename({self._resample_dim: self._dim})
+
+    def reduce(self, func, dim=None, keep_attrs=False, **kwargs):
+        """Reduce the items in this group by applying `func` along the
+        pre-defined resampling dimension.
+
+        Note that `dim` is by default here and ignored if passed by the user;
+        this ensures compatibility with the existing reduce interface.
+
+        Parameters
+        ----------
+        func : function
+            Function which can be called in the form
+            `func(x, axis=axis, **kwargs)` to return the result of collapsing an
+            np.ndarray over an integer valued axis.
+        keep_attrs : bool, optional
+            If True, the datasets's attributes (`attrs`) will be copied from
+            the original object to the new one.  If False (default), the new
+            object will be returned without attributes.
+        **kwargs : dict
+            Additional keyword arguments passed on to `func`.
+
+        Returns
+        -------
+        reduced : Array
+            Array with summarized data and the indicated dimension(s)
+            removed.
+        """
+        return super(DatasetResample, self).reduce(
+            func, self._dim, keep_attrs, **kwargs)
+
+    def _interpolate(self, kind='linear'):
+        """Apply scipy.interpolate.interp1d along resampling dimension."""
+        from .dataset import Dataset
+        from .variable import Variable
+
+        old_times = self._obj[self._dim].astype(float)
+        new_times = self._full_index.values.astype(float)
+
+        data_vars = OrderedDict()
+        coords = OrderedDict()
+
+        # Apply the interpolation to each DataArray in our original Dataset
+        for name, variable in self._obj.variables.items():
+            if name in self._obj.coords:
+                if name == self._dim:
+                    coords[self._dim] = self._full_index
+                elif self._dim not in variable.dims:
+                    coords[name] = variable
+                else:
+                    if isinstance(variable.data, dask_array_type):
+                        raise TypeError('dask arrays not supported yet in '
+                                        'resample.interpolate()')
```

Comment:
Let's make this error message more descriptive here: https://github.com/pydata/xarray/issues/1566